Add support of multiple hash types per object.

klauspost opened the pull request:
This is a basic framework that will allow multiple hash types across filesystems.

The `fs.Fs` can support the `fs.HashedFs` interface to indicate that the files on it has support for one or more hash types.

This allows us to see if we have overlapping hash type(s) between filesystems.

In the `fs.Object`, the `Md5sum() (string, error)` function is replaced by a `Hash(r fs.HashType) (string, error)` function that returns a specific hash type. I considered a more complex request type, but I would expect this to fulfill our current needs.

Still missing:

* [x] Replace all instances of `Md5Sum()` in filesystems.
* [x] Find common hash types.
* [x] ~~Maybe disable automatic hash calculation on slow systems.~~
* [x] Tests

Future work:
* Add an optional local db with known hashes of local files with size/timestamp/possibly other information to ensure validity.

Related to issues #198, #157, #265 (Syncthing uses SHA256 and Murmur3).

klauspost commented:
Ok, this should be enough for a base implementation.

@ncw, you are welcome to review the changes when you have the time. I do currently not have access to any FS with SHA-1 support, so if you are able to test that, would help greatly.

ncw commented:
I have commented on the commit.

An excellent piece of work :-)

I'll try the code next and see what happens...

ncw commented:
This seems to have broken Swift/s3/Google Cloud storage/Hubic/Drive all like this
```
rclone/fs$ go test -remote TestSwift:
--- FAIL: TestSyncAfterChangingContentsOnly (0.85s)
	fstest.go:66: potato: md5 hash incorrect - expecting "e4cb6955d9106df6263c45fcfc10f163" got "100defcf18c42a1e0dc42a789b107cd2"
--- FAIL: TestSyncAfterRemovingAFileAndAddingAFileDryRun (0.61s)
	fstest.go:66: potato: md5 hash incorrect - expecting "e4cb6955d9106df6263c45fcfc10f163" got "100defcf18c42a1e0dc42a789b107cd2"
```

axtux commented:
As I understand, you want rclone to calculate hashes of every local files on each launch (even when no checksum flag is set ?). That would be a bad idea and actually, that's why I changed my backup program to rclone.

My situation is a good example :
I'm making a backup everyday from my big local files (about 4To) to Swift using a Core 2 Duo laptop. Only changes are uploaded and it is usually done within 30 minutes. Hash calculation on these files couldn't even be done within 24h. With hashes cache, it would be long only for the first launch but I think it is an undesirable overhead.

klauspost commented:
>Hash calculation on these files couldn't even be done within 24h.

Hash calculation is done while the objects are being uploaded right now (even without this PR). If the destination supports hashing, the destination is checked. This functionality is kept as is. The only change is that SHA1 is also being calculated, so SHA-1 destination can check the hash.

The hash cache will not change any existing functionality. The only difference is that if the hash is calculated for a file, we store it, so if the hash is needed some time in the future we don't keep to recalculate. In effect this would allow you to use hashes (`--checksum`) for your backup without recalculating hashes for all your unmodified files.

axtux commented:
Thanks for clarifying. I suppose hash calculation is good to check that the file has not been corrupted during transfer. This could explain this issue #177.

The overhead should be noticeable on good upload connections with weak CPUs. Maybe 2 hashes could be ok. But the problem will be more relevant when more hashes will be added. At some point, you should consider changing this behaviour to calculate only one hash which is compatible with remote provider.

klauspost commented:
@ncw - I have implemented all the things you mentioned.

Please review, but do not merge yet. GIT messed up when I rebased, so there is some junk in there. I will fix it up, or send a new PR.

ncw commented:
I've merged the rebased version #292 so I'm closing this.
